### PR TITLE
JSDK-2255: Implement different levels of network quality score reports for local and remote participants. 

### PR DESCRIPTION
**Pull request by syerrapragada**
This PR attempts to implement different levels of network quality score reports for local and remote participants. 

@manjeshbhargav Added integration test.

### CODE REVIEW

**Comment by katranci** [comment]
Hey @syerrapragada 👋What kind of information are we going to get with different levels of network quality verbosity? Would be great to know in advance to that we can start working on the UI/UX.

**Comment by syerrapragada** [comment]
> Hey @syerrapragada 👋What kind of information are we going to get with different levels of network quality verbosity? Would be great to know in advance to that we can start working on the UI/UX.

Hi @katranci, Please take a look at the [NetworkQualityStats](https://github.com/twilio/twilio-video.js/pull/586/files#diff-0328c21afbcd8a695969dbbb10a78069) class. 

There will be following levels of verbosity that can be configured for NetworkQuality API:
**Local**:
- Level 1 (default): This is the same as today. Reports NetworkQualityLevel on LocalParticipant.
- Level 2: LocalParticipant will also have NetworkQualityStats populated with Audio and Video levels based on which the Overall NetworkQualityLevel is calculated.
- Level 3: LocalParticipant will also have detailed NetworkQualityStats populated with send/recv stats for audio and video based on which the Overall NetworkQualityLevel is calculated.

**Remote**:
- Level0 (default): No scores for RemoteParticipant
- Level 1: Reports NetworkQualityLevel on RemoteParticipant.
- Level 2: RemoteParticipant will also have NetworkQualityStats populated with Audio and Video stats based on which the Overall NetworkQualityLevel is calculated.
- Level 3: RemoteParticipant will also have detailed NetworkQualityStats populated with send/recv stats for audio and video based on which the NetworkQualityLevel is calculated.

**Comment by katranci** [comment]
Thanks @syerrapragada, that's very exciting! If level 3 is selected, does the  `networkQualityLevelChanged` event get emitted as soon as one of the quality indicators change?

File: lib/connect.js [hunk 1]
```diff
@@ -289,7 +289,7 @@ function connect(token, options) {
  *   as a hint for variable bitrate codecs, but will not take effect for fixed
  *   bitrate codecs
  * @property {?string} [name=null] - Set to connect to a {@link Room} by name
- * @property {boolean} [networkQuality=false] - Whether to enable the Network
+ * @property {boolean|NetworkQualityConfiguration} [networkQuality=false] - Whether to enable the Network
```
Comment (manjeshbhargav) [inline lib/connect.js]: @syerrapragada Can you update the description here to describe the config object?

File: lib/signaling/v2/localparticipant.js [hunk 2]
```diff
@@ -171,6 +176,16 @@ class LocalParticipantV2 extends LocalParticipantSignaling {
     }
     return publication;
   }
+
+  /**
+   * Updates the verbosity of network quality information.
+   * @param {NetworkQualityConfiguration} config
+   * @returns {void}
+   */
+  setNetworkQualityConfiguration(config) {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/localparticipant.js]: @syerrapragada Can you make this more robust? You can use the pattern used in `lib/encodingparameters.js` and create a class called `NetworkQualityConfigurationImpl` which emits a "changed" event.

File: lib/signaling/v2/networkqualitysignaling.js [hunk 3]
```diff
@@ -17,19 +18,45 @@ const AsyncVar = require('../../util/asyncvar');
  */
 
 /**
- * @interface SendAndReceiveLevels
- * @deprecated - the decomposed levels are only used for debugging and will be
- *   removed as soon as we are confident in our implementation
+ * @interface LatencyStats
```
[the remark this answers is not in the document]
Comment (syerrapragada) [reply lib/signaling/v2/networkqualitysignaling.js]: These are not public facing interfaces but JSON Objects returned by the server.

File: lib/signaling/v2/networkqualitysignaling.js [hunk 4]
```diff
@@ -165,11 +220,14 @@ class NetworkQualitySignaling extends EventEmitter {
 
 /**
  * @param {NetworkQualityInputs} networkQualityInputs
+ * @param {NetworkQualityConfigurationImpl} networkQualityConfiguration
```
Comment (manjeshbhargav) [inline lib/signaling/v2/networkqualitysignaling.js]: @syerrapragada 
```js
/**
 * @typedef {object} NetworkQualityReportLevels
 * @param {number} reportLevel
 * @param {number} remoteReportLevel
 */

/**
 * @param {NetworkQualityReportLevels} networkQualityReportLevels
 */
```

File: lib/signaling/v2/networkqualitysignaling.js [hunk 5]
```diff
@@ -165,11 +220,14 @@ class NetworkQualitySignaling extends EventEmitter {
...
  * @returns {object} message
  */
-function createNetworkQualityInputsMessage(networkQualityInputs) {
+function createNetworkQualityInputsMessage(networkQualityInputs, networkQualityConfiguration) {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/networkqualitysignaling.js]: @syerrapragada 
```js
function createNetworkQualityInputsMessage(..., networkQualityReportLevels) {
  ...
}
```

File: lib/signaling/v2/networkqualitysignaling.js [hunk 6]
```diff
@@ -165,11 +220,14 @@ class NetworkQualitySignaling extends EventEmitter {
...
   return Object.assign({
```
Comment (manjeshbhargav) [inline lib/signaling/v2/networkqualitysignaling.js]: @syerrapragada 
```js
return Object.assign(
  { type: 'network_quality' },
  networkQualityInputs,
  networkQualityReportLevels);
```

File: lib/signaling/v2/room.js [hunk 7]
```diff
@@ -53,6 +53,9 @@ class RoomV2 extends RoomSignaling {
         value: null,
         writable: true
       },
+      _networkQualityConfiguration: {
+        value: localParticipant._networkQualityConfiguration
```
Comment (manjeshbhargav) [inline lib/signaling/v2/room.js]: @syerrapragada If this is how you access the `NetworkQualityConfiguration`, then can you make it a public member of `LocalParticipantV2`?

File: lib/stats/networkqualitysendorrecvstats.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+const NetworkQualityBandwidthStats = require('./networkqualitybandwidthstats');
+const NetworkQualityFractionLostStats = require('./networkqualityfractionloststats');
+const NetworkQualityLatencyStats = require('./networkqualitylatencystats');
+
+/**
+ * Send or recv network quality statistics.
```
[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply lib/stats/networkqualitysendorrecvstats.js]: @syerrapragada Nitpick:
> Send or **receive** network quality statistics.

?

File: scripts/docs.js [hunk 9]
```diff
@@ -78,6 +88,16 @@ const privateConstructors = [
   'LocalTrackStats',
   'LocalAudioTrackStats',
   'LocalVideoTrackStats',
+  'NetworkQualityAudioStats',
+  'NetworkQualityBandwidthStats',
+  'NetworkQualityFractionLostStats',
+  'NetworkQualityLatencyStats',
+  'NetworkQualityMediaStatsq',
```
Comment (manjeshbhargav) [inline scripts/docs.js]: @syerrapragada Typo - extra **q** at the end?

File: test/integration/spec/localparticipant.js [hunk 10]
```diff
@@ -1047,4 +1047,114 @@ describe('LocalParticipant', function() {
       assert(!error);
     });
   });
+
+  (defaults.topology === 'group' ? describe : describe.skip)('"networkQualityLevelChanged" event', () => {
```
Comment (manjeshbhargav) [inline test/integration/spec/localparticipant.js]: @syerrapragada Can you make this:
```js
defaults.topology !== 'peer-to-peer'
```
to cover small group Rooms?

File: lib/networkqualityconfiguration.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        : defaultLevel;
+      if (this[participant] !== newValue) {
+        this[participant] = newValue;
+        shouldEmitChanged = true;
+      }
+      return shouldEmitChanged;
+    }, false);
+
+    if (shouldEmitChanged) {
+      this.emit('changed');
```
Comment (manjeshbhargav) [inline lib/networkqualityconfiguration.js]: @syerrapragada I think you can get rid of this event.

File: lib/connect.js [hunk 12]
```diff
@@ -375,8 +390,37 @@ const LogLevel = {
   off: 'off'
 };
 
-function createLocalParticipant(signaling, log, encodingParameters, options, localTracks) {
-  const localParticipantSignaling = signaling.createLocalParticipantSignaling(encodingParameters);
+/**
+ * Specifies verbosity of network quality information.
+ * @enum {number}
+ */
```
Comment (manjeshbhargav) [inline lib/connect.js]: @syerrapragada Nitpick: add extra line?

File: lib/signaling/v2/networkqualitysignaling.js [hunk 13]
```diff
@@ -63,11 +90,18 @@ class NetworkQualitySignaling extends EventEmitter {
...
+        value: new Map(),
+        writable: true
+      },
       _mediaSignalingTransport: {
         value: mediaSignalingTransport
       },
       _networkQualityInputs: {
         value: new AsyncVar()
+      },
+      _networkQualityConfiguration: {
```
[the remark this answers is not in the document]
Comment (syerrapragada) [reply lib/signaling/v2/networkqualitysignaling.js]: I like it better. Thank you.

File: lib/connect.js (hunk 12 above)
Comment (syerrapragada) [reply lib/connect.js]: Extra line after `*/`?

File: lib/networkqualityconfiguration.js [hunk 14]
```diff
@@ -46,30 +45,16 @@ class NetworkQualityConfigurationImpl extends EventEmitter {
...
+    [
       ['local', 1, 3, DEFAULT_NQ_LEVEL_LOCAL],
       ['remote', 0, 3, DEFAULT_NQ_LEVEL_REMOTE]
-    ].reduce((shouldEmitChanged, [participant, min, max, defaultLevel]) => {
-      const newValue = typeof networkQualityConfiguration[participant] === 'number'
-        && inRange(networkQualityConfiguration[participant], min, max)
-        ? networkQualityConfiguration[participant]
+    ].forEach(([localOrRemote, min, max, defaultLevel]) => {
+      this[localOrRemote] = typeof networkQualityConfiguration[localOrRemote] === 'number'
+          && inRange(networkQualityConfiguration[localOrRemote], min, max)
```
Comment (manjeshbhargav) [inline lib/networkqualityconfiguration.js]: @syerrapragada Nitpick: Remove the extra two spaces at the start of this line?

File: lib/connect.js (hunk 12 above)
Comment (manjeshbhargav) [reply lib/connect.js]: @syerrapragada Yes.

**Comment by manjeshbhargav** [comment]
Reopening in #603 against `master-chrome-planb`.

**Pull request closed without merging**